$OpenBSD: patch-libnetutil_netutil_cc,v 1.4 2014/02/10 17:43:15 sthen Exp $

chunks dealing with DLT_LINUX_SLL are to cope with non-bundled libpcap (not
currently used). should probably be #ifdef DLT_LINUX_SLL and go upstream?

chunks dealing with iovec: ???

--- libnetutil/netutil.cc.orig	Mon Jul 29 00:08:48 2013
+++ libnetutil/netutil.cc	Wed Oct 16 22:00:54 2013
@@ -3124,7 +3124,6 @@ static int route_dst_netlink(const struct sockaddr_sto
                              const struct sockaddr_storage *spoofss) {
   struct sockaddr_nl snl;
   struct msghdr msg;
-  struct iovec iov;
   struct nlmsghdr *nlmsg;
   struct rtmsg *rtmsg;
   struct rtattr *rtattr;
@@ -3176,22 +3175,14 @@ static int route_dst_netlink(const struct sockaddr_sto
     add_rtattr_addr(nlmsg, &rtattr, &len, RTA_SRC, spoofss, intf_index);
   }
 
-  iov.iov_base = nlmsg;
-  iov.iov_len = nlmsg->nlmsg_len;
-
   memset(&msg, 0, sizeof(msg));
   msg.msg_name = &snl;
   msg.msg_namelen = sizeof(snl);
-  msg.msg_iov = &iov;
-  msg.msg_iovlen = 1;
 
   rc = sendmsg(fd, &msg, 0);
   if (rc == -1)
     netutil_fatal("%s: cannot sendmsg: %s", __func__, strerror(errno));
 
-  iov.iov_base = buf;
-  iov.iov_len = sizeof(buf);
-
   len = recvmsg(fd, &msg, 0);
   if (len <= 0)
     netutil_fatal("%s: cannot recvmsg: %s", __func__, strerror(errno));
@@ -3822,7 +3813,6 @@ static const unsigned char *add_exthdr_ancillary(struc
 static int send_ipv6_ip(const struct sockaddr_in6 *dst,
   const unsigned char *packet, size_t packetlen) {
   struct msghdr msg;
-  struct iovec iov;
 
   const unsigned char *end;
   struct ip6_hdr *hdr;
@@ -3838,8 +3828,6 @@ static int send_ipv6_ip(const struct sockaddr_in6 *dst
   /* Set up sendmsg data structure. iov is filled in below. */
   msg.msg_name = (void *) dst;
   msg.msg_namelen = sizeof(*dst);
-  msg.msg_iov = &iov;
-  msg.msg_iovlen = 1;
   msg.msg_control = NULL;
   msg.msg_controllen = 0;
   msg.msg_flags = 0;
@@ -3898,8 +3886,6 @@ static int send_ipv6_ip(const struct sockaddr_in6 *dst
   }
 
   assert(packet <= end);
-  iov.iov_base = (unsigned char *) packet;
-  iov.iov_len = end - packet;
 
   n = sendmsg(sd, &msg, 0);
   if (n == -1)
@@ -4068,9 +4054,6 @@ int datalink_offset(int datalink)
 {
   if (datalink == DLT_EN10MB)
     return ETH_HDR_LEN;
-  else if (datalink == DLT_LINUX_SLL)
-    /* The datalink type is Linux "cooked" sockets. See pcap-linktype(7). */
-    return 16;
   else
     return -1;
 }
@@ -4107,7 +4090,7 @@ static int read_reply_pcap(pcap_t *pd, long to_usec,
     netutil_fatal("Cannot obtain datalink information: %s", pcap_geterr(pd));
   ioffset = datalink_offset(*datalink);
   if (ioffset < 0)
-    netutil_fatal("datalink_offset failed for type %d (DLT_EN10MB = %d, DLT_LINUX_SLL = %d)", *datalink, DLT_EN10MB, DLT_LINUX_SLL);
+    netutil_fatal("datalink_offset failed for type %d (DLT_EN10MB = %d)", *datalink, DLT_EN10MB);
   *offset = (unsigned int) ioffset;
 
   if (to_usec > 0) {
@@ -4207,10 +4190,6 @@ static bool accept_arp(const unsigned char *p, const s
 
   if (datalink == DLT_EN10MB) {
     return ntohs(*((u16 *) (p + 12))) == ETH_TYPE_ARP;
-  } else if (datalink == DLT_LINUX_SLL) {
-    return ntohs(*((u16 *) (p + 2))) == ARPHRD_ETHER && /* sll_hatype */
-      ntohs(*((u16 *) (p + 4))) == 6 && /* sll_halen */
-      ntohs(*((u16 *) (p + 14))) == ETH_TYPE_ARP; /* sll_protocol */
   } else {
     return false;
   }
